src/main/java/io/pinecone/PineconeDataPlaneClient.java

```diff
@@ -0,0 +1,222 @@
+            throw new PineconeValidationException("BlockingStub cannot be null.");
+        }
+
+        this.blockingStub = blockingStub;
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values) {
+        return upsert(id, values, null, null, null, null);
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values,
+                                 String namespace) {
+        return upsert(id, values, null, null, null, namespace);
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values,
+                                 List<Long> sparseIndices,
+                                 List<Float> sparseValues,
+                                 com.google.protobuf.Struct metadata,
+                                 String namespace) {
+        UpsertRequest.Builder upsertRequest = UpsertRequest.newBuilder();
+
+        if (id == null || id.isEmpty() || values == null || values.isEmpty()) {
+            throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both id and values are provided.");
+        }
+
+        Vector.Builder vectorBuilder = Vector.newBuilder()
+                .setId(id)
+                .addAllValues(values);
+
+        if (sparseIndices != null && sparseValues != null && sparseIndices.size() != sparseValues.size()) {
+            throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both sparse indices and values are of the same length.");
+        }
+
+        if (sparseIndices != null && !sparseIndices.isEmpty()) {
+            vectorBuilder.setSparseValues(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        if (metadata != null) {
+            vectorBuilder.setMetadata(metadata);
+        }
+
+        upsertRequest.addVectors(vectorBuilder.build());
+
+        if (namespace != null && !namespace.isEmpty()) {
+            upsertRequest.setNamespace(namespace);
+        }
+
+        return blockingStub.upsert(upsertRequest.build());
+    }
+
+    public QueryResponse query(String id,
+                               String namespace,
+                               int topK,
+                               Struct filter,
+                               boolean includeValues,
+                               boolean includeMetadata,
+                               List<Float> vectors,
+                               List<Long> sparseIndices,
+                               List<Float> sparseValues) {
+        QueryRequest.Builder queryRequest = QueryRequest.newBuilder();
+
+        if (id != null && !id.isEmpty() && vectors != null && !vectors.isEmpty()) {
+            throw new PineconeValidationException("Invalid Query Request. Both id and vectors cannot be present");
```

Comment:
Will update this to:
Invalid Query Request. Both Vector ID and vectors cannot be present.